libs/widget-react/src/lib/CowSwapWidget.tsx

```diff
@@ -1,37 +1,73 @@
-import { useEffect, useRef } from 'react'
+import { useEffect, useRef, useState } from 'react'
 
-import { cowSwapWidget, CowSwapWidgetParams, EthereumProvider, UpdateWidgetCallback } from '@cowprotocol/widget-lib'
+import {
+  createCowSwapWidget,
+  CowSwapWidgetParams,
+  EthereumProvider,
+  CowSwapWidgetHandler,
+} from '@cowprotocol/widget-lib'
 import type { CowEventListeners } from '@cowprotocol/events'
 
 export interface CowSwapWidgetProps {
-  params: CowSwapWidgetParams
-  listeners?: CowEventListeners
+  params: Omit<CowSwapWidgetParams, 'provider'>
   provider?: EthereumProvider
+  listeners?: CowEventListeners
 }
 
-export function CowSwapWidget({ params, provider }: CowSwapWidgetProps) {
-  const providerRef = useRef<EthereumProvider | null>()
-  const iframeContainerRef = useRef<HTMLDivElement>(null)
-  const updateWidgetRef = useRef<UpdateWidgetCallback | null>(null)
+export function CowSwapWidget({ params, provider, listeners }: CowSwapWidgetProps) {
+  const paramsRef = useRef<Omit<CowSwapWidgetParams, 'provider'> | null>(null)
+  const providerRef = useRef<EthereumProvider | null>(provider ?? null)
+  const listenersRef = useRef<CowEventListeners | undefined>(listeners)
+
+  const containerRef = useRef<HTMLDivElement>(null)
+  const widgetHandlerRef = useRef<CowSwapWidgetHandler | null>(null)
 
   useEffect(() => {
-    if (!iframeContainerRef.current) return
+    if (
+      !containerRef.current ||
+      paramsRef.current === params ||
+      JSON.stringify(paramsRef.current) === JSON.stringify(params)
```

Comment:
Why not just compare the sintringified version?
What are the changes that the string version is the different while the obj version is not?

Reply:
i will remove. It was there since i first tried to compare by equality `paramsRef.current === params` but it was not working, new object instances were created with the same params... 

being there could be argued as an optimization to not stringify if the reference don't change, but i don't think is good to do this optimizations, and code clarity and less code is best. So removing it

Reply:
57c196c83ff378b11b57b6ff1d46e9495a8bc7f9